File: scalechain-util/src/test/scala/io/scalechain/util/TimeBasedCacheSpec.scala
package io.scalechain.util

import java.io.File
import java.util.concurrent.TimeUnit

import org.scalatest.{BeforeAndAfterEach, FlatSpec, Matchers, Suite}

class TimeBasedCacheSpec extends FlatSpec with BeforeAndAfterEach with Matchers {

  this: Suite =>

  val testPath = new File("./target/unittests-IncompleteBlockCacheSpec-storage/")

  var cache : TimeBasedCache[Integer, String] = null

  val CACHE_KEEP_MILLISECONDS = 10

  override def beforeEach() {
    super.beforeEach()

    cache = new TimeBasedCache[Integer, String](CACHE_KEEP_MILLISECONDS, TimeUnit.MILLISECONDS)
  }

  override def afterEach() {
    super.afterEach()

    cache = null
  }

  "getBlock" should "return None if no block/transaction was added" in {
    cache.get(1) shouldBe None
  }

  "getBlock" should "return an IncompleteBlock if a signing transaction was added" in {
    cache.put(1, "foo")

    cache.get(1) shouldBe Some("foo")
  }

  "getBlock" should "return None if a signing transaction was added but expired" in {
    cache.put(1, "foo")

    Thread.sleep(CACHE_KEEP_MILLISECONDS + 10)

    cache.get(1) shouldBe None
  }


}

File: scalechain-util/src/test/kotlin/io/scalechain/util/TimeBasedCacheSpec.kt
package io.scalechain.util

import io.kotlintest.*
import io.kotlintest.matchers.*
import io.kotlintest.specs.*
import org.junit.runner.RunWith

import java.io.File
import java.util.concurrent.TimeUnit


@RunWith(KTestJUnitRunner::class)
class TimeBasedCacheSpec : FlatSpec(), Matchers {

    val testPath = File("./target/unittests-IncompleteBlockCacheSpec-storage/")

    var cache : TimeBasedCache<Int, String>? = null

    val CACHE_KEEP_MILLISECONDS = 50L

    override fun beforeEach() {
        // set-up code
        cache = TimeBasedCache<Int, String>(CACHE_KEEP_MILLISECONDS, TimeUnit.MILLISECONDS)

        super.beforeEach()
    }

    override fun afterEach() {
        super.afterEach()

        // tear-down code
        //
        cache = null
    }

    init {
        "getBlock" should "return None if no block/transaction was added" {
            cache?.get(1) shouldBe null
        }

        "getBlock" should "return an IncompleteBlock if a signing transaction was added" {
            cache?.put(1, "foo")

            cache?.get(1) shouldBe "foo"
        }

        "getBlock" should "return None if a signing transaction was added but expired" {
            cache?.put(1, "foo")

            Thread.sleep(CACHE_KEEP_MILLISECONDS + 10)

            cache?.get(1) shouldBe null
        }
    }
}